torch/autograd/functional.py

```diff
@@ -0,0 +1,717 @@
+        raise RuntimeError("v is a tuple of invalid length: should be {} but got {}.".format(len(other), len(v)))
+
+    for idx, (el_v, el_other) in enumerate(zip(v, other)):
+        if el_v.size() != el_other.size():
+            raise RuntimeError("Entry {} in v has invalid size: should be {} but got {}.".format(
+                               idx, el_other.size(), el_v.size()))
+
+def _check_requires_grad(inputs, input_type, strict):
+    # Used to make all the necessary checks to raise nice errors in strict mode.
+    if not strict:
+        return
+
+    if input_type not in ["outputs", "grad_inputs", "jacobian", "hessian"]:
+        raise RuntimeError("Invalid input_type to _check_requires_grad")
+    for i, inp in enumerate(inputs):
+        if inp is None:
+            # This can only be reached for grad_inputs.
+            raise RuntimeError("The output of the user-provided function is independent of input {}."
+                               " This is not allowed in strict mode.".format(i))
+        if not inp.requires_grad:
+            if input_type == "hessian":
+                raise RuntimeError("The hessian of the user-provided function with respect to input {}"
+                                   " is independent of the input. This is not allowed in strict mode."
+                                   " You should ensure that your function is thrice differentiable and that"
+                                   " the hessian depends on the inputs.".format(i))
+            elif input_type == "jacobian":
+                raise RuntimeError("The jacobian of the user-provided function with respect to input {}"
+                                   " is independent of the input. This is not allowed in strict mode."
+                                   " You should ensure that your function is twice differentiable and that"
+                                   " the jacobian depends on the inputs (this would be violated by a linear"
+                                   " function).".format(i))
+            elif input_type == "grad_inputs":
+                raise RuntimeError("The gradient with respect to input {} is independent of the inputs of the"
+                                   " user-provided function. This is not allowed in strict mode.".format(i))
+            else:
+                raise RuntimeError("Output {} of the user-provided function does not require gradients."
+                                   " The outputs must be computed in a differentiable manner from the input"
+                                   " when running in strict mode.".format(i))
+
+def _autograd_grad(outputs, inputs, grad_outputs=None, create_graph=False, retain_graph=None):
+    # Version of autograd.grad that accepts `None` in outputs and do not compute gradients for them.
+    # This has the extra constraint that inputs has to be a tuple
+    if torch.is_tensor(outputs):
+        outputs = (outputs,)
+        grad_outputs = (grad_outputs,)
```

Comment:
do you need to check that outputs and grad_outputs have the same structure?

Reply:
Yes I simplified this: we assume we get a tuple of inputs and assert accordingly.